moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi
+
+import com.squareup.moshi.internal.canonicalize
+import com.squareup.moshi.internal.hasNullable
+import com.squareup.moshi.internal.jsonAnnotations
+import com.squareup.moshi.internal.toStringWithAnnotations
+import java.io.IOException
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+internal class AdapterMethodsFactory(
+  private val toAdapters: List<AdapterMethod>,
+  private val fromAdapters: List<AdapterMethod>
+) : JsonAdapter.Factory {
+  override fun create(
+    type: Type,
+    annotations: Set<Annotation>,
+    moshi: Moshi
+  ): JsonAdapter<*>? {
+    val toAdapter = get(toAdapters, type, annotations)
+    val fromAdapter = get(fromAdapters, type, annotations)
+    if (toAdapter == null && fromAdapter == null) return null
+
+    val delegate: JsonAdapter<Any>? = if (toAdapter == null || fromAdapter == null) {
```

Comment:
The nullability for the rest of this function is sort of all over the place. Can you please clean this up (avoiding `!!` mainly)

Reply:
https://github.com/square/moshi/pull/1467/commits/3e6e9bda3372997dff18ac5d1567f6b80707074e#diff-e8a15c8ca843b45b53cf5d1b2097710c70dd4b5a9399c074e7fa2b379a772d47R61